### ISSUE TITLE
Excessively long argument values in parameterized display names should be truncated

### ISSUE THREAD

**Issue by vlsi** [junit-team/junit5#2358]
JUnit provides a nice way to configure test names, however, the names might become too long unexpectedly.

Here's a case for JUnit4:

https://github.com/pgjdbc/pgjdbc/blob/aa190cce2c27ac5f5d4fa33f5d156f8bf04cbe0c/pgjdbc/src/test/java/org/postgresql/core/UTF8EncodingTest.java#L29-L33

```java
  @Parameterized.Parameters(name = "string={1}, encoding={0}")
  public static Iterable<Object[]> data() {
    final StringBuilder reallyLongString = new StringBuilder(1024 * 1024);
    for (int i = 0; i < 185000; ++i) {
      reallyLongString.append(i);
    }
```

I guess the same is relevant for JUnit5 as well.

It would be great if JUnit could detect such cases, and it could trim the sub-parts before it produces a single multi-megabyte test name.

For instance, in the case like above JUnit could produce something like `string=012345678910111213...truncated...sha1=12312312123, encoding=utf-8`

It turns out the long test names consume disk space as Gradle records all the test outputs, and it does make `remote build cache` slower, as the build cache has to keep all the long test names (so it impacts `cache space consumption`, `cache store latency` and `cache retrieval latency`).

Of course, users can truncate test names on their own, however, I believe it would be better if JUnit prevented "unexpectedly long test name" issue automatically.

In other words:

1) If use is using placeholder-like pattern, then JUnit should trim all the parameters.
2) If the user returns test name via programmatic API (I'm not sure if that exists at all, but I assume it does :) ), then JUnit could truncate the full string.
3) There should probably be a configuration option for the default limit of the parameter value.

WDYT?

Relevant Gradle Build Cache issue: https://github.com/gradle/gradle/issues/13608#issuecomment-660874776

**Comment by marcphilipp** [comment]
I think instead of truncating we could impose a limit on display names that could be changed via a configuration parameter. This way, unexpectedly long display names would not go unnoticed. WDYT?

**Comment by marcphilipp** [comment]
Tentatively slated for 5.7 M2 solely for the purpose of _team discussion_.

**Comment by vlsi** [comment]
In my case (see ticket description), the test is parameterized. Some of the instances are short, and some of them are long:

```java
  public static Iterable<Object[]> data() {
    final StringBuilder reallyLongString = new StringBuilder(1024 * 1024);
    for (int i = 0; i < 185000; ++i) {
      reallyLongString.append(i);
    }

    final List<String> strings = new ArrayList<String>(150);
    strings.add("short simple"); // <-- this is short
    strings.add("longer but still not really all that long"); // <-- this is short as well
    strings.add(reallyLongString.toString());  // <-- this is LONG
```

That is why I think it is better to implement truncation (with a hash for keeping test names unique) rather than "fail on long test names".

Well, an alternative way could be to enhance `string={1}` like `string={1 | truncateTo(50)}` or something like that. However, I don't like making `{N}` syntax complicated as it would be hard to write.

**Comment by marcphilipp** [comment]
This concrete use case could only be solved by a feature specific to parameterized tests, e.g. a new annotation attribute of `@ParameterizedTest`, since you would want to see the remaining parameters when the name is being truncated. Is the checksum really necessary?

**Comment by vlsi** [comment]
Note: I suggest to truncate parameter-by-parameter first, and truncate full
name only in case it overflows.

For example, there might be 512 characters per parameter, and 5120 overall
name limit.

That is why the case could be solved automatically, and other small
parameters would be visible.

Checksum is for keeping test names unique, so test reports are actionable

>

**Comment by marcphilipp** [comment]
> For example, there might be 512 characters per parameter

How would you like to configure that? As `@ParameterizedTest(maxParameterDisplayNameLength = 512)` or sth. like that?

> Checksum is for keeping test names unique, so test reports are actionable, and it is easier to understand which one works and which one fails

Each invocation already has a unique ID and a "legacy reporting name" that's unique. There shouldn't be a strict requirement for display names to be unique except to avoid confusion... 😉

**Comment by vlsi** [comment]
>How would you like to configure that? As @ParameterizedTest(maxParameterDisplayNameLength = 512) or sth. like that?

`maxParameterDisplayNameLength = 512` would be OK, however, I'm not sure if it is too much exposure for a public API :-)

The following would work as well:

`junit.jupiter.execution.param.max_param_in_display_name.length = 512`
`junit.jupiter.execution.param.max_display_name.length = 5120`

For instance, I'm always adding the following properties, so configuring (e.g. in case override needed) parameter truncation would be ok:

```
junit.jupiter.execution.parallel.enabled=true
junit.jupiter.execution.parallel.mode.default=concurrent
junit.jupiter.execution.timeout.default=5 m
```

>There shouldn't be a strict requirement for display names to be unique except to avoid confusion... 

I know JUnit supports tests with equivalent display names, however, let us please generate human-friendly names ;-)

**Comment by marcphilipp** [comment]
**Team decision:** Let's introduce a `junit.jupiter.params.displayname.argument.maxlength` configuration parameter to truncate the values included in display names of parameterized test invocations with `…` but omit the checksum and the overall display name truncation for now.

**Comment by marcphilipp** [comment]
@vlsi Do you have time to work on a PR for this?

**Comment by vlsi** [comment]
Is there a reason to keep the truncated values ambiguous?

**Comment by marcphilipp** [comment]
They're not ambiguous by default since we prepend the invocation index.

**Comment by marcphilipp** [comment]
I'm starting on this now.
